### Nonlinear Problems in Finance (MATH-GA.2045-001), New York University, Fall 2019
# Homework III

### Due Date: 11:55 PM Sunday, November 10, 2019


You should turn in the notebook at NYU Classes website.

Please comment your code properly.

Before you turn in the notebook, press the "Run all cells" button in the toolbar, and make sure all the calculation results and graphs are produced correctly in a reasonable time frame, and then save the notebook.

# Uncertain Mortality Model vs American Options

** Re-insurance Deals**

In this assignment, we study the *Uncertain Mortality Model* for the pricing of reinsurance deals. For the sake of simplicity, in this assignment we only consider one type of default: the risk of death.

We consider the following reinsurance product:

- At maturity, if the insurance subscriber is alive, the issuer delivers a put on the underlying $X$

$$
u^\textrm{mat}( x) = (K_\textrm{mat} - x)_+.
$$

- At the time of death, if it is before the maturity, the issuer delivers an exit payoff, typically another put on the underlying $X$.

$$
u^D(t, x) = (K_D - x)_+.
$$

- The subscriber pays a constant fee $α\Delta t$ at every time step until death or the maturity or the product

- The insurance sells a large number of these contracts to subscribers. We assume that the times of death of the subscribers $\tau^D$ are independent, and identically distributed, and also independent of the underlying's stock price.

- We assume that the underlying's risk neutral price dynamics is the Black-Scholes model with zero interest rate/repo/dividend yield

$$
dX_t = \sigma X_t d W_t.
$$

** The Insurers' Approach and Risk-Neutral Pricing**

This contract shows two types of risk: the times of death of the subscribers and the changes in the price of the underlying. 

In this case, the issuer can apply the insurer's approach to the risk of death times, i.e., the law of large numbers. The more people buy the contract, the less risk.

Choosing a risk-neutral measure under which the death times $\tau^D$ have the same distribution as under the historical probability measure is equivalent to applying the arbitrage-pricing approach to the financial risk insurer's rule on the risk of death. The price of the contract is then

$$
u(t, x) = \mathbb{E}^\mathbb{Q}_{t, x} \left[u^\textrm{mat}(X_T) \mathbb{1}_{\tau^D \geq T} -\int_t^T\alpha\mathbb{1}_{\tau^D>s}ds+ u^D(\tau^D, X_{\tau^D}) \mathbb{1}_{\tau^D < T} \right].
$$

** Deterministic Death Rate **

If the death intensity is a deterministic function $\lambda_t^D$ (i.e. $\tau^D$ has an exponential distribution with time-dependent intensity $\lambda_t^D$), then we have seen that $u$ is the solution to the linear PDE

$$
\left\{\begin{array}{l}
\partial_t u + \frac{1}{2} \sigma^2 x^2 \partial_x^2 u -\alpha + \lambda_t^D \cdot (u^D - u) = 0,\\
u(T, \cdot) \equiv u^\textrm{mat}.
\end{array}\right.
$$

** Uncertain Mortality Model **

If the death rate is uncertain, we assume that it is adapted (i.e., it does not look into the future) and belongs to a moving corridor
$\left[\underline{\lambda}_t, \overline{\lambda}_t\right]$. The most conservative way to price the contract is to compute the (financially) worst death rate process $\lambda_t^D$ as being chosen so as to maximize the value of the contract. The resulting HJB equation is

$$
\left\{\begin{array}{l}
\partial_t u + \frac{1}{2} \sigma^2 x^2 \partial^2_x u -\alpha + \Lambda^D(t, u^D - u) \cdot (u^D - u) = 0,\\
u(T, \cdot) \equiv u^\textrm{mat},
\end{array}\right.
$$
where the function $\Lambda^D$ is defined by
$$
\Lambda^D (t, y) = \left\{\begin{array}{l}
    \overline{\lambda}^D_t \quad \textrm{if} \ y \geq 0, \\
    \underline{\lambda}^D_t \quad \textrm{otherwise.}
\end{array}\right.
$$

## Link with $1$-BSDE and Numerical Schemes

From the Pardoux-Peng theorem, we know that the solution $u(0, x)$ can be represented as the solution $Y_0^x$ to the $1$-BSDE

$$
dY_t = -f(t, X_t, Y_t, Z_t) \, dt + Z_t \, dW_t,
$$

with the terminal condition $Y_T = u^\textrm{mat} (X_T)$, where $X_0 = x$ and

$$
f(t, x, y, z) = -\alpha + \Lambda^D(t, u^D(t, x) - y) \cdot (u^D(t, x) - y).
$$


## BSDE Discretization

### Explicit Euler Schemes

\begin{align*}
Y_{t_n} =& \ u^{\text{mat}}\left(X_{t_n}\right)\\
Y_{t_{i - 1}} =& \ \mathbb{E}^\mathbb{Q}_{i - 1} [ Y_{t_i} ] -\alpha\Delta t_i + \Lambda^D \left(t_{i - 1}, u^D(t_{i - 1} , X_{t_{i - 1}}) - \mathbb{E}^\mathbb{Q}_{i - 1} [ Y_{t_i} ]\right) \cdot \left( u^D(t_{i - 1}, X_{t_{i - 1}}) - \mathbb{E}^\mathbb{Q}_{i - 1} [ Y_{t_i} ] \right) \Delta t_i\\
=& -\alpha\Delta t_i + \left( \mathbb{E}^\mathbb{Q}_{i - 1} [ Y_{t_i} ] \left( 1 - \overline{\lambda}^D \Delta t_i \right) +
u^D(t_{i - 1}, X_{t_{i - 1}}) \overline{\lambda}^D \Delta t_i \right) \mathbb{1}_{u^D\left(t_{i - 1}, X_{t_{i - 1}}\right) \geq \mathbb{E}_{i-1}[Y_{t_i}]}\\
& + \left( \mathbb{E}^\mathbb{Q}_{i - 1} [ Y_{t_i} ] \left( 1 - \underline{\lambda}^D \Delta t_i \right) +
u^D(t_{i - 1}, X_{t_{i - 1}}) \underline{\lambda}^D \Delta t_i \right) \mathbb{1}_{u^D\left(t_{i - 1}, X_{t_{i - 1}}\right) < \mathbb{E}_{i-1}[Y_{t_i}]}.
\end{align*}

### Implicit Euler Scheme

\begin{align*}
Y_{t_n} =& \ u^{\text{mat}}\left(X_{t_n}\right)\\
Y_{t_{i - 1}} =& \ \mathbb{E}^\mathbb{Q}_{i - 1} [Y_{t_i}] -\alpha\Delta t_i+ \Lambda^D \left(t_{i - 1}, u^D(t_{i - 1}, X_{t_{i - 1}}) - Y_{t_{i - 1}}\right) \cdot \left( u^D(t_{i - 1}, X_{t_{i - 1}}) - Y_{t_{i - 1}} \right) \Delta t_i.
\end{align*}

The implicit scheme involves $Y_{t_{i - 1}}$ on both sides of the equation, which generally requires a root-finding routine to find $Y_{t_{i - 1}}$. However in this specific case, it can be shown that 

\begin{equation}u^D(t_{i-1}, X_{t_{i-1}}) \geq Y_{t_{i-1}}\quad\text{if and only if}\quad u^D(t_{i-1}, X_{t_{i-1}}) \geq \mathbb{E}^{\mathbb{Q}}_{i-1}\left[Y_{t_i}\right]-\alpha\Delta t_i.\end{equation}

Thus

\begin{split}
Y_{t_{i - 1}} =& \frac{1}{1 + \overline{\lambda}^D \Delta t_i} \left( \mathbb{E}^\mathbb{Q}_{i - 1} [ Y_{t_i} ] - \alpha\Delta t_i +
u^D(t_{i - 1}, X_{t_{i - 1}}) \overline{\lambda}^D \Delta t_i \right) \mathbb{1}_{u^D\left(t_{i - 1}, X_{t_{i - 1}}\right) \geq \mathbb{E}_{i-1}[Y_{t_i}]-\alpha\Delta t_i}\\
& + \frac{1}{1 + \underline{\lambda}^D \Delta t_i} \left( \mathbb{E}^\mathbb{Q}_{i - 1} [ Y_{t_i} ] -\alpha\Delta t_i +
u^D(t_{i - 1}, X_{t_{i - 1}}) \underline{\lambda}^D \Delta t_i \right) \mathbb{1}_{u^D(t_{i - 1}, X_{t_{i - 1}}) < \mathbb{E}_{i-1}[Y_{t_i}]-\alpha\Delta t_i}.
\end{split}



In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import norm

In [2]:
def blackscholes_mc(S=100, vol=0.2, r=0, q=0, ts=np.linspace(0, 1, 13), npaths=10):
    """Generate Monte-Carlo paths in Black-Scholes model.

    Parameters
    ----------
    S: scalar
        The spot price of the underlying security.
    vol: scalar
        The implied Black-Scholes volatility.
    r: scalar
        The annualized risk-free interest rate, continuously compounded.
    q: scalar
        The annualized continuous dividend yield.
    ts: array_like
        The time steps of the simualtion
    npaths: int
        the number of paths to simulate

    Returns
    -------
    paths: ndarray
        The Monte-Carlo paths.
    """
    #np.random.seed(0)
    nsteps = len(ts) - 1
    ts = np.asfarray(ts)[:, np.newaxis]
    W = np.cumsum(np.vstack((np.zeros((1, npaths), dtype=np.float),
                             np.random.randn(nsteps, npaths) * np.sqrt(np.diff(ts, axis=0)))),
                  axis=0)
    paths = np.exp(-0.5*vol**2*ts + vol*W)*S*np.exp((r-q)*ts)
    return paths

def blackscholes_price(K, T, S, vol, r=0, q=0, callput='call'):
    """Compute the call/put option price in the Black-Scholes model
    
    Parameters
    ----------
    K: scalar or array_like
        The strike of the option.
    T: scalar or array_like
        The maturity of the option, expressed in years (e.g. 0.25 for 3-month and 2 for 2 years)
    S: scalar or array_like
        The current price of the underlying asset.
    vol: scalar or array_like
        The implied Black-Scholes volatility.
    r: scalar or array_like
        The annualized risk-free interest rate, continuously compounded.
    q: scalar or array_like
        The annualized continuous dividend yield.
    callput: str
        Must be either 'call' or 'put'.

    Returns
    -------
    price: scalar or array_like
        The price of the option.

    Examples
    --------
    >>> blackscholes_price(95, 0.25, 100, 0.2, r=0.05, callput='put')
    1.5342604771222823
    """
    F = S*np.exp((r-q)*T)
    v = np.sqrt(vol**2*T)
    d1 = np.log(F/K)/v + 0.5*v
    d2 = d1 - v
    try:
        opttype = {'call':1, 'put':-1}[callput.lower()]
    except:
        raise ValueError('The value of callput must be either "call" or "put".')
    price = opttype*(F*norm.cdf(opttype*d1)-K*norm.cdf(opttype*d2))*np.exp(-r*T)
    return price


def pwlinear_basis(xknots):
    """Basis that represent a piecewise linear function with given knots"""
    fs = [lambda x: np.ones_like(x, dtype=np.float), lambda x: x-xknots[0]]
    fs.extend([lambda x, a=xknots[i]: np.maximum(x-a, 0) for i in range(len(xknots))])
    return fs


def pwlinear_fit(xdata, ydata, xknots):
    """Fit a piecewise linear function with xknots to xdata and ydata"""
    fs = pwlinear_basis(xknots)
    A = np.vstack([f(xdata) for f in fs]).T
    ps = np.linalg.lstsq(A, ydata, rcond=None)[0]
    return ps, fs

<b>Note.</b> To be more specific about the payouts of the product, we make the following assumptions:

- We assume that all transactions occur at the end of the month, i.e. the insurance subsriber is expected to pay a fee of $\alpha/12$ at the end of each month; if the subscriber dies in the middle of the month, the death payment will be made at the end of that month. 

- If the death occurs at the maturity, we assume that the death comes first and the death payment $u^D$ will be made.

## Assignment III

<b style="color:darkorange">Question 1</b> (Deterministic Mortality Rate Model)

We assume that the underlying's risk neutral price dynamics is the Black-Scholes model with zero interest rate/repo/dividend yield

$$dX_t = \sigma X_t d W_t,\quad X_0=100,\quad\sigma=0.3$$

We shall price the reinsurance deal using two Monte Carlo methods explained below.

<b>(a).</b> (Direct simulation of death times) Implement a Monte Carlo simulation by directly simulating death time $\tau^D$ for each path to estimate the quantity

$$u(t, x) = \mathbb{E}^\mathbb{Q}_{t, x} \left[u^\textrm{mat}(X_T) \mathbb{1}_{\tau^D \geq T} - \int_t^T\alpha \mathbb{1}_{\tau^D> s}ds
+ u^D(\tau^D, X_{\tau^D}) \mathbb{1}_{\tau^D < T} \right],$$

when $T=10$, $K_{\text{mat}}=90$, $K_D=100$, $\alpha=3$, and $\tau^D$ has an exponential distribution of constant intensity $\lambda^D=0.025$. Use monthly time steps.

<b>(b).</b> (Averaing over death events) Using the Feynman-Kac formula, derive a stochastic representation of the linear PDE 

$$\left\{\begin{array}{l}\partial_t u + \frac{1}{2}\sigma^2 x^2 \partial_x^2 u - \alpha + \lambda^D \cdot (u^D - u) = 0\\
        u(T, x) \equiv u^\textrm{mat}(x)\end{array}\right.$$
        
that does not involve simulating death times $\tau^D$. Explain how one can get this stochastic representation from the one given in (a). Implement the corresponding Monte Carlo simulation to estimate $u(0,X_0)$.

In [3]:
# note that lambda is a reserved keyword in python, so we use lambd instead
def exponential_samples(n, lambd):
    """Generate n samples from exponential distribution with rate lambd"""
    return -np.log(np.random.rand(n))/lambd

### Question 1.a

In [4]:
def u(x, type = "M"):
    """
    Examples:
    >>> u(0, 'M')
    90
    >>> u(0, 'D')
    100
    >>> u(0, 'd')
    Error type!
    """
    if type == "M":
        return np.maximum(K_mat - x, 0)
    elif type == "D":
        return np.maximum(K_D - x, 0)
    else:
        raise Exception("Error type!")

In [5]:
# parameter setting
X0 = 100
sig = 0.3

T = 10
K_mat = 90.0
K_D = 100.0
alpha = 3
lambda_d = 0.025

ts = np.linspace(0, T, int(np.round(T*12))+1)

In [34]:
npaths = 50000
paths = blackscholes_mc(X0, sig, 0, 0, ts, npaths)

tau_D = exponential_samples(npaths, lambda_d)

payoffs = []

for i in range(len(ts)):
    if i == len(ts) - 1:
        payoffs.append(np.sum(u(paths[-1], "M") - ts[i]*alpha))
    else:
        ind = tau_D <= ts[i]
        payoffs.append(np.sum(u(paths[i][ind], "D") - ts[i]*alpha))
        paths = paths[:,~ind]
        tau_D = tau_D[~ind]
        if paths == []:
            break

print(np.sum(payoffs) / npaths)

c:\users\tiany\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:16: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  app.launch_new_instance()


1.8898669512728927


### Question 1.b

Explanation: Please see the PDF attached (Page 1-3)

In [57]:
# parameter setting
X0 = 100
sig = 0.3

T = 10
K_mat = 90.0
K_D = 100.0
alpha = 3
lambda_d = 0.025

ts = np.linspace(0, T, int(np.round(T*12))+1)

In [58]:
def calculator_MC(step, p):
    if (step == (len(ts) - 1)):
        return u(paths[step], 'M')
    else:
        return u(paths[step + 1], 'D') * p + calculator_MC(step + 1, p) * (1 - p) - del_t*alpha

In [59]:
#calculate increment length, probability of death p, probability of no death (1-p)
del_t = np.diff(ts)[0]
p = 1 - np.exp(-(lambda_d * del_t))
np.mean(calculator_MC(0, p))

2.007840242644662

<b style="color:darkorange">Question 2</b> (Uncertain Mortality Rate Model - BSDE)

Assume that $\lambda^D_t \in \left[\underline{\lambda}^D, \overline{\lambda}^D \right]$, where $\underline{\lambda}^D=0.005$ and $\overline{\lambda}^D=0.04$.

<b>(a).</b> Implement the implicit Euler scheme for BSDE. Use piecewise-linear fit with 10 knots to estimate the conditional expectation $\mathbb{E}^{\mathbb{Q}}_{i-1}\left[Y_{t_i}\right]$. Use the numpy function percentile to define the minimum and maximum knots. Use the percentiles 1 and 99 percent. Use 100,000 paths.

<b>(b).</b> Run an independent new simuation. Use the estimates of $\mathbb{E}^{\mathbb{Q}}_{i-1}\left[Y_{t_i}\right]$ obtained in (a) to get a lower bound estimate.

<b>(c).</b> How does the price compare with the price in the model with deterministic mortality rate $\underline{\lambda}^D$? with deterministic mortality rate $\overline{\lambda}^D$? Check numerically.

### Question 2.a

In [60]:
# parameter setting
X0 = 100
sig = 0.3

T = 10
K_mat = 90.0
K_D = 100.0
alpha = 3

lambda_lower_d = 0.005
lambda_upper_d = 0.04

ts = np.linspace(0, T, int(np.round(T*12))+1)

In [61]:
npaths = 100000
paths = blackscholes_mc(X0, sig, 0, 0, ts, npaths)

V = u(paths[-1], "M")

fits = []

for i in range(len(ts)-2, -1, -1):
    del_t = ts[i+1] - ts[i]
    xknots = np.linspace(np.percentile(paths[i], 1), np.percentile(paths[i], 99), 10)
    ps, fs = pwlinear_fit(paths[i], V, xknots)
    est = sum([f(paths[i])*p for (f, p) in zip(fs, ps)])
    booleans = u(paths[i],'D') >= (est - (alpha*del_t))
    V_upper = (1 / ((lambda_upper_d * del_t) + 1)) * (est -(alpha * del_t) + (u(paths[i], 'D') * lambda_upper_d * del_t))
    V_lower = (1 / ((lambda_lower_d * del_t) + 1)) * (est -(alpha * del_t) + (u(paths[i], 'D') * lambda_lower_d * del_t))
    V = booleans * V_upper + (1 - booleans) * V_lower
    fits.append((fs,ps))
np.mean(V)   

3.2869761798141743

### Question 2.b

In [62]:
# parameter setting
X0 = 100
sig = 0.3

T = 10
K_mat = 90.0
K_D = 100.0
alpha = 3

lambda_lower_d = 0.005
lambda_upper_d = 0.04

ts = np.linspace(0, T, int(np.round(T*12))+1)

In [63]:
npaths = 100000
paths = blackscholes_mc(X0, sig, 0, 0, ts, npaths)

plain_fits = fits[::-1]

V = u(paths[-1],'M')

for i in range(len(ts)-2, -1, -1):
    del_t = ts[i+1] - ts[i]
    xknots = np.linspace(np.percentile(paths[i], 1), np.percentile(paths[i], 99), 10)
    ps, fs = pwlinear_fit(paths[i], V, xknots)
    oldest = sum([f(paths[i])*p for (f, p) in zip(plain_fits[i][0], plain_fits[i][1])])
    booleans = u(paths[i],'D') >= (oldest - (alpha*del_t))
    V_upper = (1 / ((lambda_upper_d * del_t) + 1)) * (V -(alpha * del_t) + (u(paths[i],'D') * lambda_upper_d * del_t))
    V_lower = (1 / ((lambda_lower_d * del_t) + 1)) * (V -(alpha * del_t) + (u(paths[i],'D') * lambda_lower_d * del_t))
    V = booleans * V_upper + (1 - booleans) * V_lower
    
np.mean(V)

3.152694509925549

### Question 2.c

In [66]:
# parameter setting
X0 = 100
sig = 0.3

T = 10
K_mat = 90.0
K_D = 100.0
alpha = 3

lambda_lower_d = 0.005
lambda_upper_d = 0.04

ts = np.linspace(0, T, int(np.round(T*12))+1)
npaths = 100000
paths = blackscholes_mc(X0, sig, 0, 0, ts, npaths)
del_t = ts[1] - ts[0]
def price(Lambda_d):
    
    p = 1 - np.exp(-(Lambda_d * del_t))
    return np.mean(calculator_MC(0, p))

print("Price given by Lambda (" + str(lambda_lower_d) + ") is ", price(lambda_lower_d))
print("Price given by Lambda (" + str(lambda_upper_d) + ") is ", price(lambda_upper_d))

Price given by Lambda (0.005) is  0.3300706705833617
Price given by Lambda (0.04) is  3.12341690462595


Comments:

By the lower bound BSDE, the reinsurance price is around 3.16 dollars (Question 2.b). The price given by the deterministic lambda 0.04 (using function from Question 1.b) is roughly 3.12 dollars. The price given by the deterministic lambda 0.005 (using function from Question 1.b) is roughly 0.33 dollars.

We find that the lower bound BSDE price actually is higher than the prices given by deterministic lambda_d. We may explain this by using HJB:

Using uncertain death rate actually gives the sup_bound of all the possible prices is 'selected' at a certain period. That's why, more possibly, the reinsurance can be hedged by using the higher price resulting from the HJB. That's why the lower bound BSDE price is larger than the deterministic two.

Also, the price given by the deterministic lambda 0.04 is also close to the HJB price. This is because the when the death rate is high and the reinsurance deal buyer has more demands which leads to higher price.

<b style="color:darkorange">Question 3</b> (Uncertain Mortality Rate Model - Longstaff and Schwartz) In addition to the BSDE simulation, the reinsurance deals can be priced by adapting the Longstaff-Schwartz algorithm. An implementation of the Longstaff-Schwartz algorithm is provided below for your convenience.

The same lower and upper bounds of $\lambda_t^D$ are used as Question 2.

<b>(a).</b> In the example code, we use European put price (the strike is taken to be the average of $K_{\text{mat}}$ and $K_D$) and constant as basis functions to estimate the sum of future payoffs. Modify the code by using the piecewise-linear fit to estimate the sum of future payoffs.

<b>(b).</b> Compare the BSDE scheme with the Longstaff-Schwartz-like scheme in terms of variance. For each method run the two-step procedure (regression and indedenpent pricing) 20 times to estimate variance. Use 5000 paths for regression and 50000 for indepedent pricing.

<b>(c).</b> What is approximately the value of the fee $\alpha$ that makes the deal costless at inception?

### Question 3.a

In [67]:
# Longstaff-Schwartz algorithm for Uncertain Mortality Rate Model
S = 100
vol = 0.3
T = 10
ts = np.linspace(0, T, int(np.round(T*12))+1)
alpha = 3
lambda_min = 0.005
lambda_max = 0.04

K_mat = 90.0
K_D = 100.0

def u_mat(x):
    return np.maximum(K_mat - x, 0)

def u_D(x):
    return np.maximum(K_D - x, 0)


# first Monte Carlo run to estimate the value function by regressions
npaths = 5000
paths = blackscholes_mc(S=S, vol=vol, r=0, q=0, ts=ts, npaths=npaths)
#betas = np.zeros((len(ts), 2), dtype=np.float)
V = u_mat(paths[-1])
lambd = np.where(u_D(paths[-1]) >= V, lambda_max, lambda_min)
K = 0.5*(K_mat+K_D)
fits = []
for i in range(len(ts)-2, -1, -1): # iterate to 0 rather than 1
    dt = ts[i+1]-ts[i]
    p = 1-np.exp(-lambd*dt)
    V = p*u_D(paths[i+1]) + (1-p)*V - alpha*dt
    #Z = blackscholes_price(K, ts[-1]-ts[i], paths[i], vol, callput='put')
    #A = np.vstack((np.ones_like(Z), Z)).T
    #betas[i] = np.linalg.lstsq(A, V, rcond=None)[0]
    knots = np.linspace(np.percentile(paths[i], 1), np.percentile(paths[i], 99), 10)
    #calculate the weights for each basis function, vector B, using least squares minimization
    #also returns the piecewise functions 
    ps, fs = pwlinear_fit(paths[i], V, knots)
    est = sum([f(paths[i])*p for (f, p) in zip(fs, ps)])
    fits.append((fs,ps))
    lambd = np.where(u_D(paths[i]) >= est, lambda_max, lambda_min)
    
plain_fits = fits[::-1]
# independent simulation to obtain a lower bound
npaths = 100000
paths = blackscholes_mc(S=S, vol=vol, r=0, q=0, ts=ts, npaths=npaths)
V = u_mat(paths[-1])
lambd = np.where(u_D(paths[-1]) >= V, lambda_max, lambda_min)
K = 0.5*(K_mat+K_D)
for i in range(len(ts)-2, -1, -1):
    dt = ts[i+1]-ts[i]
    p = 1-np.exp(-lambd*dt)
    V = p*u_D(paths[i+1]) + (1-p)*V - alpha*dt
    #Z = blackscholes_price(K, ts[-1]-ts[i], paths[i], vol, callput='put')
    oldest = sum([f(paths[i])*p for (f, p) in zip(plain_fits[i][0], plain_fits[i][1])])
    lambd = np.where(u_D(paths[i]) >= oldest, lambda_max, lambda_min)
np.mean(V)

3.225678244477901

### Question 3.b

In [111]:
S = 100
vol = 0.3
T = 10
ts = np.linspace(0, T, int(np.round(T*12))+1)
alpha = 3
lambda_min = 0.005
lambda_max = 0.04

K_mat = 90.0
K_D = 100.0

def BSDE(alpha):
    npaths = 5000
    paths = blackscholes_mc(S=S, vol=vol, r=0, q=0, ts=ts, npaths=npaths)

    V = u(paths[-1], "M")

    #matrix to store regression informations
    fits = []

    for i in range(len(ts)-2, -1, -1):
        del_t = ts[i+1] - ts[i]
        xknots = np.linspace(np.percentile(paths[i], 1), np.percentile(paths[i], 99), 10)
        ps, fs = pwlinear_fit(paths[i], V, xknots)

        est = sum([f(paths[i])*p for (f, p) in zip(fs, ps)])
        booleans = u(paths[i],'D') >= (est - (alpha*del_t))
        V_upper = (1 / ((lambda_upper_d * del_t) + 1)) * (est -(alpha * del_t) + (u(paths[i], 'D') * lambda_upper_d * del_t))
        V_lower = (1 / ((lambda_lower_d * del_t) + 1)) * (est -(alpha * del_t) + (u(paths[i], 'D') * lambda_lower_d * del_t))
        V = booleans * V_upper + (1 - booleans) * V_lower
        fits.append((fs,ps))
    npaths = 50000
    paths = blackscholes_mc(S=S, vol=vol, r=0, q=0, ts=ts, npaths=npaths)

    plain_fits = fits[::-1]

    V = u(paths[-1],'M')

    for i in range(len(ts)-2, -1, -1):
        del_t = ts[i+1] - ts[i]
        xknots = np.linspace(np.percentile(paths[i], 1), np.percentile(paths[i], 99), 10)
        ps, fs = pwlinear_fit(paths[i], V, xknots)
        oldest = sum([f(paths[i])*p for (f, p) in zip(plain_fits[i][0], plain_fits[i][1])])
        booleans = u(paths[i],'D') >= (oldest - (alpha*del_t))
        V_upper = (1 / ((lambda_upper_d * del_t) + 1)) * (V -(alpha * del_t) + (u(paths[i],'D') * lambda_upper_d * del_t))
        V_lower = (1 / ((lambda_lower_d * del_t) + 1)) * (V -(alpha * del_t) + (u(paths[i],'D') * lambda_lower_d * del_t))
        V = booleans * V_upper + (1 - booleans) * V_lower
    return np.mean(V) 

def Longstaff_Schwartz(alpha):
    # first Monte Carlo run to estimate the value function by regressions
    npaths = 5000
    paths = blackscholes_mc(S=S, vol=vol, r=0, q=0, ts=ts, npaths=npaths)
    #betas = np.zeros((len(ts), 2), dtype=np.float)
    V = u_mat(paths[-1])
    lambd = np.where(u_D(paths[-1]) >= V, lambda_max, lambda_min)
    K = 0.5*(K_mat+K_D)
    fits = []
    for i in range(len(ts)-2, -1, -1): # iterate to 0 rather than 1
        dt = ts[i+1]-ts[i]
        p = 1-np.exp(-lambd*dt)
        V = p*u_D(paths[i+1]) + (1-p)*V - alpha*dt
        #Z = blackscholes_price(K, ts[-1]-ts[i], paths[i], vol, callput='put')
        #A = np.vstack((np.ones_like(Z), Z)).T
        #betas[i] = np.linalg.lstsq(A, V, rcond=None)[0]
        knots = np.linspace(np.percentile(paths[i], 1), np.percentile(paths[i], 99), 10)
        #calculate the weights for each basis function, vector B, using least squares minimization
        #also returns the piecewise functions 
        ps, fs = pwlinear_fit(paths[i], V, knots)
        est = sum([f(paths[i])*p for (f, p) in zip(fs, ps)])
        fits.append((fs,ps))
        lambd = np.where(u_D(paths[i]) >= est, lambda_max, lambda_min)

    plain_fits = fits[::-1]
    # independent simulation to obtain a lower bound
    npaths = 50000
    paths = blackscholes_mc(S=S, vol=vol, r=0, q=0, ts=ts, npaths=npaths)
    V = u_mat(paths[-1])
    lambd = np.where(u_D(paths[-1]) >= V, lambda_max, lambda_min)
    K = 0.5*(K_mat+K_D)
    for i in range(len(ts)-2, -1, -1):
        dt = ts[i+1]-ts[i]
        p = 1-np.exp(-lambd*dt)
        V = p*u_D(paths[i+1]) + (1-p)*V - alpha*dt
        #Z = blackscholes_price(K, ts[-1]-ts[i], paths[i], vol, callput='put')
        oldest = sum([f(paths[i])*p for (f, p) in zip(plain_fits[i][0], plain_fits[i][1])])
        lambd = np.where(u_D(paths[i]) >= oldest, lambda_max, lambda_min)
    return np.mean(V)

In [115]:
BSDE_logs = [BSDE(3) for i in range(20)]
Longstaff_Schwartz_logs = [Longstaff_Schwartz(3) for i in range(20)]

print("Population Variance of the BSDE (biased) = {}".format(np.var(BSDE_logs, ddof=0)))
print("Population Variance of the Longstaff_Schwartz (biased) = {}".format(np.var(Longstaff_Schwartz_logs, ddof=0)))
print("\n")

print("Sample Variance of the BSDE (Unbiased) = {}".format(np.var(BSDE_logs, ddof=1)))
print("Sample Variance of the Longstaff_Schwartz (Unbiased) = {}".format(np.var(Longstaff_Schwartz_logs, ddof=1)))
print("\n")

print("Mean of the BSDE = {}".format(np.mean(BSDE_logs)))
print("Mean of Longstaff_Schwartz = {}".format(np.mean(Longstaff_Schwartz_logs)))

Population Variance of the BSDE (biased) = 0.09975756994158426
Population Variance of the Longstaff_Schwartz (biased) = 0.01597418263017781


Sample Variance of the BSDE (Unbiased) = 0.10500796835956237
Sample Variance of the Longstaff_Schwartz (Unbiased) = 0.016814929084397696


Mean of the BSDE = 3.140112102923218
Mean of Longstaff_Schwartz = 3.2164625880867277


In [78]:
def find_root(func, l, h, tolerance = 1e-3, max_iter = 100):
    """
    Using binary search to get the root of the algorithm "func"
    """
    def stable_result(func, alpha):
        return np.mean([func(alpha) for i in range(50)])
    iter_ = 0
    while True:
        print("Iteration epoch: ", iter_)
        print("The lower bound of the searching area is ", l)
        print("The upper bound of the searching area is ", h)
        print("\n")
        m = (l + h)/2
        if stable_result(func, m) < 0:
            h = m
        else:
            l = m
        iter_ += 1
        if np.abs(h-l) < tolerance:
            break
        if iter_ == max_iter:
            print("The root is not found in " + str(iter_) + " iterations")
    return (l + m)/2

In [79]:
print("The value of the 𝛼 found by the Longstaff Schwartz algorithm that makes the deal costless at inception: ", find_root(Longstaff_Schwartz,3.3,3.4))
print("\n")
print("The value of the 𝛼 found by BSDE that makes the deal costless at inception: ", find_root(BSDE,3.3,3.4))

Iteration epoch:  0
The lower bound of the searching area is  3.3
The upper bound of the searching area is  3.4


Iteration epoch:  1
The lower bound of the searching area is  3.3499999999999996
The upper bound of the searching area is  3.4


Iteration epoch:  2
The lower bound of the searching area is  3.375
The upper bound of the searching area is  3.4


Iteration epoch:  3
The lower bound of the searching area is  3.375
The upper bound of the searching area is  3.3875


Iteration epoch:  4
The lower bound of the searching area is  3.375
The upper bound of the searching area is  3.38125


Iteration epoch:  5
The lower bound of the searching area is  3.378125
The upper bound of the searching area is  3.38125


Iteration epoch:  6
The lower bound of the searching area is  3.3796875
The upper bound of the searching area is  3.38125


The value of the 𝛼 found by the Longstaff Schwartz algorithm that makes the deal costless at inception:  3.3804687500000004


Iteration epoch:  0
The lower

<b style="color:darkorange">Question 4.</b> Set $K_D=0$, and all the other parameters remain the smae, in particualr $\alpha=3$. Then the default event is equivalent to lapse (with no mortality payoff). It has been observed that insurance subscribers do not lapse optimally, which explains the use of the uncertain lapse model with minimum and maximum lapse rates $\underline{\lambda}$, $\overline{\lambda}$. 

<b>(a).</b> If insurance subscribers were to lapse optimally, i.e., exercise optimally their option to lapse, how would you price the contract?

<b>(b).</b> Explain how you can get this price (from optimal exercise) in the uncertain lapse model. Reuse the code provided in Question 3 to show your result (you may have to change some parameters in the code).

<b>(c).</b> Implement Longstaff-Schwartz algorithm to price the American option with payoff $u^D=0$ (Always perform a second indepedent run to get a clean lower bound price). Compare the price from uncertain lapse model in part (b) with the American price.

### Question 4.a
Because $K_D = 0$, $u^D = max\{K_D - 0, 0\}\equiv 0 $

Actually, the insurance were to lapse optimally when $0 = u^D > u$ because under this situation to lapse is better than a to continue on with the insurance. When $0 = u^D \leq u$, there is no lapse, which indicates $\lambda^D_s = 0$.

To put it mathematically, under this situation, we have:
$$\left\{\begin{array}{l}\partial_t u + \frac{1}{2}\sigma^2 x^2 \partial_x^2 u - \alpha = 0\\
        u(T, x) = u^\textrm{mat}(x)\end{array}\right.$$
        
Put it into variational inequality form, we have:
$$max\{\partial_t u + \frac{1}{2}\sigma^2 x^2 \partial_x^2 u - \alpha, u^D - u\} = 0$$

We can just solve this variational inequality by numerical schemes (like what we do to solve American Option numerically).

From this intuition, we can also know the 4.b should almost share the same result of 4.c because $\lambda^D$ ranges from 0 to $\infty$ which means 4.b shares the same exercise oppotunity to the American Option discribed in 4.c.

### Question 4.b
Because the insurance were to lapse optimally only when $0 = u^D > u$ because under this situation to lapse is better than a to continue on with the insurance. We can just price reinsurance use the Longstaff_Schwartz(alpha) function that I wrote in Question 3.b. We can use $\bar{\lambda}^D = \infty$ to represemt the lapse which means the first arriving time in infinity. Also, as I said in Question 4.a, when $0 = u^D \leq u$, there is no lapse, which indicates $\underline{\lambda}^D = 0$. 

In [107]:
from math import inf as INF
S = 100
vol = 0.3
T = 10
ts = np.linspace(0, T, int(np.round(T*12))+1)
alpha = 3
lambda_min = 0
lambda_max = INF

K_mat = 90.0
K_D = 0

Longstaff_Schwartz(alpha)

3.6302072525461004

### Question 4.c

In [108]:
# parameter setting
X0 = 100
sig = 0.3

T = 10
K_mat = 90.0
alpha = 3

ts = np.linspace(0, T, int(np.round(T*12))+1)
del_t = ts[1] - ts[0]

In [109]:
# Longstaff_Schwartz for American Option pricing
npaths = 10000
paths = blackscholes_mc(X0, sig, 0, 0, ts, npaths)

fits = []

payoff = u(paths[-1], "M")
for i in range(len(ts)-2, -1, -1): # to from len(ts)-2 to 0
    payoff = payoff - (alpha*del_t)
    xknots = np.linspace(np.percentile(paths[i], 1), np.percentile(paths[i], 99), 10)
    ps, fs = pwlinear_fit(paths[i], payoff, xknots)
    if i < (len(ts)-2):
        contval = sum([f(paths[i])*p for (f, p) in zip(fs, ps)]) 
    else:
        contval = payoff
    exerval = (np.ones_like(paths[1,:], dtype=np.float) * 0.0) - (alpha * del_t)
    ind = exerval >= contval
    payoff[ind] = exerval[ind]
    fits.append((fs,ps))

plain_fits = fits[::-1]  
    
# second time MC
npaths = 100000
paths = blackscholes_mc(S=S, vol=vol, r=0, q=0, ts=ts, npaths=npaths)

payoff = u(paths[-1], "M")
for i in range(len(ts)-2, -1, -1): # to from len(ts)-2 to 0
    payoff = payoff - (alpha*del_t)
    if i < (len(ts)-2):
        contval = sum([f(paths[i])*p for (f, p) in zip(plain_fits[i][0], plain_fits[i][1])])
    else:
        contval = payoff
    exerval = (np.ones_like(paths[1,:], dtype=np.float) * 0.0) - (alpha * del_t)
    ind = exerval >= contval
    payoff[ind] = exerval[ind]

np.mean(payoff)    

3.6584342089487585

The result price of 4.b (using a Longstaff_Schwartz with lambda from 0 to $\infty$) is roughly 3.63 dollars.
The result low biased price in 4.c (Longstaff_Schwartz with a second indepedent run) is roughly 3.66 dollars.

As I mentioned in 4.a, intuitively speaking, the 4.b should almost share the same result of 4.c because $\lambda^D$ ranges from 0 to $\infty$ which means 4.b shares the same exercise oppotunity to the American Option discribed in 4.c.

However, there are differences between the algorithms in 4.b and 4.c. 

First, in 4.b, the lambda of $t_{i-1}$ is based on the information of $t_{i}$ which means the algorithm is actually using some further information whereas the algorithm in 4.c doesn't have the problem it always calculate the future infomation based on the past information (using regression).

Second, if people don't use a large $\bar{\lambda}^D$, there will be some errors here.